### DISCUSSION TITLE
Structuring for Inheritance/Overriding

### DISCUSSION THREAD

**Discussion by daviddyball** [cue-lang/cue#1618] [Q&A]
Hi All :wave:

I'm coming from a `kustomize` background, so correct me if my thinking isn't conducive to Cue.

I want to develop a configuration format that allows you to specify some globals (e.g. `base/` manifests in `kustomize`) and then apply patches (`overlays/*`) and was wondering the best way to do it.

I've come up with the following
```
#PatchEnvironments: "ci" | "staging" | "production"
#PatchClusters: "cluster1" | "cluster2" | "cluster3"

#EnvConfig: {
  Env: [string]: string
}

Application: "my-app": {
  #EnvConfig

  Env: {
    ITEM1: "value1"
    ITEM2: "value2"
  }
  Patch: [#PatchEnvironments]: [#PatchClusters]: 
}

Application: "my-app": Patch: ci: cluster1: Env: ITEM3: "value3"
```
Result:
```json
{
  "Env": {
    "ITEM1": "value1",
    "ITEM2": "value2"
  },
  "Patch": {
    "ci": {
      "cluster1": {
        "Env": { "ITEM3": "value3" }
      }
    }
  }
}
```
I'm now struggling to see how to condense this down to something like this:
```json
{
  "Env": {
    "ITEM1": "value1",
    "ITEM2": "value2",
    "ITEM3": "value3"
  }
}
```
I have seen the `kubernetes` tutorial, but I'm struggling to wrap my head around the various layers involved (e.g. that tutorial uses `deployment` rather than `#Deployment`... am I using `#` definitions incorrectly above?)

Any pointers are greatly appreciated.

Cheers,
David.

**Comment by eonpatapon** [comment]
Definitions are fine because with them you can easily catch typos in field names and constrain which fields are mandatory.

To manage configurations on multiple envs I heavily rely on cue templates. Maybe you can translate it to your particular use case.

I define a map which contains the different envs I manage, and let's say the apps to be deployed on each env, something like that:

```cue
package apps

envs: {
	prod: {
		apps: {
			appA: #Config
			appB: #Config
		}
	}
	staging: {
		apps: {
			appA: #Config
		}
	}
}

#Config: {
	inputs: {...}
}
```

Now for each app you can define a global template like this, in it's own file:

```cue
package apps

envs: [string]: apps: ["appA"]: {
   inputs: {
     foo: "bar"
   }
}
```

Which means for any env (`string`), if there is an app `appA` defined, use the inputs `foo: "bar"`.

With this structure in place it is possible to define specific inputs on a specific env like so:

```cue
envs: prod: apps: appA: {
  inputs: bar: 42
}
```

Regexp is also possible:

```cue
envs: [=~ "^dev"]: apps: appA: {
  inputs: bar: 55
}
```

Disjunctions: 

envs: ["prod" | "staging"]: apps: appA: {
  inputs: bar: 42
}

Or alternatively use conditionals in the global template:

```cue
package apps

envs: [EnvName=string]: apps: ["appA"]: {
   inputs: {
     foo: "bar"
     if EnvName == "prod" {
       bar: 42
     }
   }
}
```

The global idea is that you have a path to the configuration of any app and this path can consist on different element that you can match to specialize or augment the app config if you need to.

Of course each app can have it's own definition, with possible fields and values, default values etc..
